### PR TITLE
Aether: Solution for Issue #177

### LINKED ISSUE #177: perf: ~2.9× throughput by switching from DeepSpeed to DDP
## TL;DR

The default distributed backend for pi05 LIBERO training (`distributed_type: DEEPSPEED, zero_stage: 2`) leaves **~65% of GPU time on the table at fixed hardware**. Comparing **same-hardware** 8×A100 configurations, switching the training script to plain DDP (`distributed_type: MULTI_GPU`) combined with two small follow-on fixes (drop an unused `gemma_expert.lm_head` parameter, default fused AdamW) gives **2.90× more samples/s**, ~3.4× faster wall-clock per optimizer step, and **~2.9× fewer GPU-hours per sample** — without changing any model code or loss.

> **Distinguishing hardware scaling from software gains.** The pre-investigation libero accelerate config had `num_processes: 4`. A user upgrading from that 4-GPU config to the new 8-GPU DDP config observes `13.8 → 83.1 samples/s` = **~6.0× end-to-end**, but roughly half of that comes from doubling the hardware (DDP scales near-linearly 1→8 GPUs on this model, see the per-rank throughput discussion below). **The software-attributable win is 2.90×.** This issue focuses on the software axis.

Root cause is a known DeepSpeed ZeRO pathology for models with many small parameter tensors (PaliGemma + Gemma expert): per-parameter Python hooks installed on every backward pass dominate step time and cannot be reduced by any communication-overlap knob. Single-rank profiling shows the model is healthy and compute-bound; the overhead only appears once DeepSpeed is in the loop.

## What we ruled out

We worked through the obvious suspects in order. Each one is conclusively *not* the cause:

1. **Dataloader starvation.** `profile_dataloader.py` (no model, no collective) ran at **597 cluster-wide samples/s** while training consumed only ~29 samples/s under ZeRO-2. `dataload_wait` per step is **1.14 ms** in the actual training loop. Despite the suggestive square-wave pattern in the `nvidia-smi` chart, the input pipeline is roughly 20× ahead of the GPUs.
2. **Per-step host/device sync (5× `gather_for_metrics(...).item()` in `update_policy`).** Measured at **3 ms / 0.1%** of step time. Real, but immaterial.
3. **NCCL transport / topology.** `nvidia-smi topo -m` shows every GPU pair connected by `NV12` (12 NVLinks via NVSwitch). NCCL init log confirms `P2P/CUMEM/read` on 24 channels. Reduce-scatter of 8 GB of bf16 grads across 8 ranks should take ~25 ms on this fabric.
4. **DeepSpeed comms not overlapping.** Adding `overlap_comm: true`, `contiguous_gradients: true`, `reduce_bucket_size: 5e8`, `allgather_bucket_size: 5e8`, `round_robin_gradients: true` to the DeepSpeed config produced **zero measurable improvement** (3071 ms → 3093 ms backward on 4-rank ZeRO-2, and similarly on 8-rank). Critical signal — if the bottleneck were comms, these would have helped.
5. **`find_unused_parameters=True` at `src/opentau/scripts/train.py:128`.** Silently ignored under DeepSpeed (accelerate wraps with the DeepSpeed engine, not DDP). Active under DDP — does add some cost there, but DDP still beats DeepSpeed by 2.64× even with it on.
6. **Eager attention / fp32 Q,K upcast / no SDPA / no torch.compile / no gradient checkpointing.** All real items but none explain the *pattern*. Forward time is ~380 ms and consistent with raw bf16 FLOPs on a ~3.4 B-param model at bs=12 on A100. The model is roughly compute-optimal in forward; problem is purely backward. SDPA and gradient checkpointing were later attempted in Phase 3; see the [follow-up comment](https://github.com/TensorAuto/OpenTau/issues/177#issuecomment-4301905944) for data confirming neither helps on this specific workload.

## What we zeroed in on

The decisive measurement was a single-GPU run with the same script and same model:

| Phase | DeepSpeed ZeRO-2 (4-rank, bs=12) | **Single GPU (bs=13)** | Delta |
|---|---:|---:|---:|
| forward | 391 ms | 376 ms | ~same |
| **bwd** | **3071 ms** | **641 ms** | **~2.4 s of pure DeepSpeed overhead** |
| total | 3468 ms | 1025 ms | |

Single-GPU `bwd` of 641 ms is exactly the expected ~1.7× forward — pure compute, healthy. The extra **~2400 ms on multi-GPU is not GPU compute and not GPU communication.** It's CPU-side Python execution of post-accumulation hooks DeepSpeed installs on every parameter tensor: each PaliGemma LayerNorm, bias, gate proj, etc. fires its own hook between backward kernels, stalling the next kernel launch. With hundreds of small parameter tensors, the host time dominates the step.

The "low GPU utilization" originally observed in `nvidia-smi` is the direct consequence: util measures *fraction of time at least one kernel is active*, so a 3 s wall-clock backward in which only ~450 ms is real kernel execution registers as ~15% util — exactly the chart we started from.

ZeRO-1 has the same hook mechanism and shows the same symptom (29.5 vs 28.7 samples/s on 8 GPUs — within noise of ZeRO-2). DDP installs gradient-bucket all-reduce hooks at *bucket* granularity rather than parameter granularity, so the host overhead is amortized over hundreds of params per call rather than paid once per param.

## Benchmark — pi05 / LIBERO / 8×A100 80GB (fixed hardware, same model)

Profile script: `src/opentau/scripts/profile_step.py` (added in PR #176). Each row uses the *max* per-rank batch size that fits in 80 GB, verified with `nvidia-smi`. Backward column = sum of `bwd + unscale_clip + optim_step + zero_grad_sched`. All 8-rank runs differ only in the `accelerate launch --config_file` argument (+ the post-PR fixes on DDP).

| Backend | bs/rank | global bs | bwd (ms) | total (ms) | **samples/s** | same-HW speedup | GPU mem |
|---|---:|---:|---:|---:|---:|---:|---:|
| ZeRO-2 8× *(tuned)* — **baseline** | 14 | 112 | 3442 | 3909 | **28.7** | **1.00×** | 79.7 GB |
| ZeRO-1 8× *(control)* | 14 | 112 | 3334 | 3799 | 29.5 | 1.03× | 79.7 GB |
| DDP 8× *(fup=true, `lm_head` still present)* | 11 | 88 | 665 | 1160 | **75.8** | **2.64×** | 78.0 GB |
| DDP 8× + drop `lm_head` *(fup=false)* | 12 | 96 | — | 1216 | 78.9 | 2.75× | 79.2 GB |
| **DDP 8× + drop `lm_head` + fused AdamW** — **post-PR** | 12 | 96 | — | 1155 | **83.1** | **2.90×** | 79.2 GB |
| Single GPU *(scaling reference)* | 13 | 13 | 705 | 1248 | 10.4 | — | 78.2 GB |

**DDP scaling from 1→8 GPUs: 83.1 / 10.4 = 7.99×** — near-ideal. The per-rank throughput on DDP 8× (83.1 / 8 = 10.4 s/s/rank) matches single-GPU, meaning the gradient all-reduce is fully overlapped with backward compute. ZeRO-2 gets 28.7 / 8 = 3.6 s/s/rank — a ~65% per-rank efficiency loss vs. single-GPU, entirely host-side.

ZeRO-2 backward actually **got worse** going from 4 to 8 GPUs (3071 ms → 3442 ms) even with overlap_comm enabled. More ranks = more coordination pressure on the same per-parameter Python pathway. Conclusive evidence the bottleneck is host-side and not bandwidth-limited.

### GPU-hour economics (1M optimizer steps, fixed 8×A100 hardware)

| Config | step time | samples/step | wall-clock | GPU-hours | per 1M samples |
|---|---:|---:|---:|---:|---:|
| Pre-PR: ZeRO-2 8× @ bs=14 | 3909 ms | 112 | 1086 h (45 d) | 8688 | 77.6 GPU-h |
| **Post-PR: DDP 8× + fused AdamW @ bs=12** | **1155 ms** | 96 | **321 h (13 d)** | **2568** | **26.8 GPU-h** |

**~2.89× cheaper per sample** (per-sample cost is the fair metric since the two configs run different batch sizes). Matches the samples/s ratio exactly, as expected.

## External references confirming this is a known DeepSpeed pathology

- **[deepspeedai/DeepSpeed#226 — "ZeRO-2 per-epoch training time significantly slower than ZeRO-1"](https://github.com/microsoft/DeepSpeed/issues/226)** — exact symptom we observed (~2× slower than ZeRO-1 with no expected memory savings), open on a model with many small parameters. Same pattern.
- **[deepspeedai/DeepSpeed#4856 — "ZeRO Stage 2 Memory Leak? How to cleanup?"](https://github.com/microsoft/DeepSpeed/issues/4856)** — discussion attributes the issue to `create_reduce_and_remove_grad_hooks`, which is the same per-parameter hook mechanism we're paying CPU time for.
- **[DeepSpeed ZeRO-3 docs](https://deepspeed.readthedocs.io/en/latest/zero3.html)** — ZeRO-3 explicitly classifies modules below a granularity threshold as "fine-grained" so they're treated as integral units, "to reduce host overhead and the separate allgather overhead introduced by hooks for fine-grained layers." This is DeepSpeed's own acknowledgement that the per-parameter-hook overhead is a real problem; ZeRO-3 has a workaround for it, ZeRO-2 does not.
- **[HuggingFace accelerate ZeRO docs](https://huggingface.co/docs/accelerate/usage_guides/deepspeed)** — explicit recommendation: *"if your model fits in DDP, pure DDP will be fastest."* Pi05 does fit in DDP on A100 80GB at bs=12, so we follow this guidance.

The reason this didn't show up in HF's own benchmarks (which often report ZeRO-2 ≈ DDP at higher batch sizes) is that those benchmarks use models with relatively few large parameter tensors (DeBERTa, T5, etc.). Pi05 is the bad case: PaliGemma + Gemma expert have hundreds of distinct small tensors (every LayerNorm, embed, gate proj, bias), each one paying the Python hook cost.

## Fix landed in PR #176

**Three production changes**:

1. **Switch the libero distributed default from DeepSpeed ZeRO-2 to plain DDP** — biggest contributor at 2.64× alone. File: `configs/libero/reproduce_pi05_libero_accelerate_config.yaml` (+ the new `configs/examples/accelerate_ddp_config.yaml` as the documented example).
2. **Drop unused `gemma_expert.lm_head`** from pi05 and pi0 — 263 M-param tensor that receives no gradient under any training config (the action expert is encoder-only; its output flows through `action_out_proj` / `da_head`, never through a token head). Lets DDP run with `find_unused_parameters=False` and frees memory to go bs=11 → bs=12. pi05_mem and pi05_continuous_state import the same `PaliGemmaWithExpertModel` class and inherit the fix transitively.
3. **Default to fused AdamW** (`torch.optim.AdamW(fused=True)`). Cuts `optimizer.step()` from 99.6 ms → 39.7 ms (60% reduction).

**Diagnostic infrastructure** also landed: `profile_step.py`, `profile_dataloader.py`, `find_unused_params.py` under `src/opentau/scripts/`. Used to reproduce the benchmarks above and to verify future policies don't regress.

**What did NOT help** and was reverted (see [follow-up comment](https://github.com/TensorAuto/OpenTau/issues/177#issuecomment-4301905944)):

- SDPA attention (FA-2 backward recompute tax ≈ forward savings at pi05's sequence length ~1500 tokens) — break-even at fixed bs, reverted.
- Gradient checkpointing (compute tax ~33% vs. ~2% amortization gain from the fixed per-step cost; net ~10% loss even at bs=37) — reverted.

These remain viable for *larger* pi05 variants (7B+, longer action horizons) where the compute/memory tradeoffs flip. Tracked as open possibilities, not near-term action items.

## Profiling artefacts and reproduction

- Profile scripts: `src/opentau/scripts/profile_step.py`, `profile_dataloader.py`, `find_unused_params.py` (PR #176).
- Post-PR benchmark command:
  ```bash
  FIND_UNUSED_PARAMS=false accelerate launch \
      --config_file configs/libero/reproduce_pi05_libero_accelerate_config.yaml \
      src/opentau/scripts/profile_step.py \
      --config_path=configs/libero/reproduce_pi05_libero.json \
      --batch_size=12
  # Expected: samples/s ≈ 83, total ≈ 1155 ms, optim_step ≈ 40 ms
  ```
- Pre-PR baseline A/B (reproducible — just point at the DeepSpeed example config):
  ```bash
  accelerate launch \
      --config_file configs/examples/accelerate_deepspeed_config.yaml \
      src/opentau/scripts/profile_step.py \
      --config_path=configs/libero/reproduce_pi05_libero.json \
      --batch_size=14
  # Expected: samples/s ≈ 28-29, total ≈ 3900 ms
  ```
- Branch: `claude/debug-gpu-utilization-WiDcZ`
- PR: #176

## Action items

- [x] **Promote DDP as the recommended default** in `docs/source/tutorials/training.rst` and `docs/source/installation.rst` (done in PR #176, commit `4929110`).
- [x] **Update the libero reproduction recipe** (`configs/libero/reproduce_pi05_libero_accelerate_config.yaml`) to DDP 8-GPU bf16 (done in PR #176, same commit).
- [x] **Fix pi05 / pi0 unused parameter** (`gemma_expert.lm_head`) so DDP can run with `find_unused_parameters=False` (done in PR #176, commit `e1c8a38`).
- [x] **Default fused AdamW** on `AdamConfig` / `AdamWConfig` (done in PR #176, commit `ed834f3`).
- [x] **Consolidate experimental accelerate configs** to a single `configs/examples/accelerate_ddp_config.yaml` (done in PR #176, commit `4929110`).
- [x] **Keep `configs/examples/accelerate_deepspeed_config.yaml`** for CI coverage and memory-constrained users, documented as a fallback (done).
- [x] **Add docs note** about the DeepSpeed/per-parameter-hook pitfall so the next person picking ZeRO for a vision-language policy A/Bs against DDP first (done in `installation.rst` and `training.rst`).
- [ ] **Follow-up PR**: audit remaining sibling policies (`pi05_continuous_state`, `pi05_mem`, `value`, `pi0`) for their own config defaults and benchmark whether they benefit from the same DDP + fused-AdamW treatment. pi0 already got the `lm_head` fix; the others inherit it transitively via the shared `PaliGemmaWithExpertModel`.
- [ ] **Deferred**: `torch.compile` on the policy (+15-20% expected). Interacts awkwardly with `.to("cpu")` calls in `modeling_pi05.py:691,716` — separate investigation.
- [ ] **Deferred**: SDPA and gradient checkpointing for larger pi05 variants. Infrastructure not yet in place; worth revisiting when we train a 7B+ policy or use longer action horizons.

cc generated by automated investigation; sourced from PR #176 and on-box benchmarks on `gpu-387` (8×A100 80GB, NV12 NVSwitch fabric).

### PR DESCRIPTION
**Pull request by jimbo782**
Automated high-integrity solution for issue #177.

Closes #177

*Applied BCOS-Compliance heuristics.*

### CODE REVIEW

**Comment by shuheng-liu** [comment]
Nice try
